### ISSUE TITLE
[BUG] Wild Card Query Using Brackets Returns No Metrics

### ISSUE THREAD

**Issue by cthiel42** [go-graphite/go-carbon#407]
**Describe the bug**
I'm attempting to use graphite's bracketed wildcard query syntax to query metrics and it appears like this doesn't completely work in go-carbon (I'm using v0.15.2) like it does in the original python based graphite as specified in their docs [here](https://graphite.readthedocs.io/en/latest/render_api.html#paths-and-wildcards). Note that all the queries I'm doing are done through Grafana, and sent to carbonapi, which is talking with go-carbon's built in carbonserver.

Here's an example of the kind of metrics I'm trying to query using the wildcard:
````
app-prd-product.rest.of.the.metric.path
app_prd_product.rest.of.the.metric.path
````

The query I'm attempting in Grafana looks like this:
`app[-_]prd[-_]product.*`
The above returns metrics on the python version of graphite, but in the go-carbon version it returns nothing as if there were no metrics found.

On a somewhat side note, we noticed that if an escape character is used before the hyphen, metrics are returned correctly. This works on both graphite and go-carbon. The query would resemble something like this:
`app[\-_]prd[\-_]product.*`

**Logs**

Here's what our logs from go-carbon look like for these types of queries:
````
INFO [access] find success 
{
  "handler": "find", 
  "url": "/metrics/find/?format=carbonapi_v3_pb", 
  "peer": "127.0.0.1:54228", 
  "carbonapi_uuid": "----", 
  "carbonzipper_uuid": "----", 
  "query": ["app[-_]prd[-_]product.{host1,host2,host3}.CpuUtilization"],
  "format": "carbonapi_v3_pb", 
  "runtime_seconds": 0.003869698, 
  "Files": 0, 
  "find_cache_enabled": false, 
  "from_cache": false, 
  "http_code": 200
}
````

And here's what our carbonapi logs look like if this is of any help to you:
````
INFO	access	request served	
{"data": 
    {"handler":"render",
     "carbonapi_uuid":"----",
     "url":"/render",
     "peer_ip":"----",
     "host":"----",
     "format":"json",
     "use_cache":true,
     "targets":["app[-_]prd[-_]product.{host1,host2,host3}.CpuUtilization"],
     "cache_timeout":60,
     "metrics":["app[-_]prd[-_]product.{host1,host2,host3}.CpuUtilization"],
     "have_non_fatal_errors":true,
     "runtime":0.004682361,
     "http_code":200,
     "carbonapi_response_size_bytes":2,
     "from":1615480923, 
     "until":1615484523,
     "max_data_points":336,
     "from_raw":"-1h",
     "until_raw":"now",
     "uri":"/render",
     "from_cache":false,
     "used_backend_cache":false,
     "request_headers":{}
    }
}
````

**Go-carbon Configuration:**
````
[common]
user = "root"
graph-prefix = "carbon.agents.{host}"
metric-endpoint = "local"
metric-interval = "1m0s"
max-cpu = 16

[whisper]
data-dir = "/opt/graphite/storage/whisper"
schemas-file = "/opt/go-carbon/storage-schemas.conf"
aggregation-file = "/opt/go-carbon/storage-aggregation.conf"
workers = 8
max-updates-per-second = 0
max-creates-per-second = 0
hard-max-creates-per-second = false
sparse-create = false
flock = true
enabled = true
hash-filenames = true
compressed = true
remove-empty-file = false

[cache]
max-size = 0
write-strategy = "noop"

[udp]
listen = ":2003"
enabled = true
buffer-size = 0

[tcp]
listen = ":2003"
enabled = true
buffer-size = 0

[pickle]
listen = ":2004"
max-message-size = 67108864
enabled = true
buffer-size = 0

[carbonlink]
listen = "127.0.0.1:7002"
enabled = true
read-timeout = "30s"

[grpc]
listen = "127.0.0.1:7003"
enabled = true

[tags]
enabled = false

[carbonserver]
listen = "127.0.0.1:8080"
enabled = true
buckets = 10
metrics-as-counters = false
read-timeout = "60s"
write-timeout = "60s"
query-cache-enabled = false
query-cache-size-mb = 2000
find-cache-enabled = false
trigram-index = false
scan-frequency = "5m0s"
trie-index = false
cache-scan = false
max-globs = 10000
fail-on-max-globs = true
max-metrics-globbed  = 100000
max-metrics-rendered = 10000
graphite-web-10-strict-mode = true
internal-stats-dir = ""
stats-percentiles = [99, 98, 95, 75, 50]

[dump]
enabled = false

[pprof]
enabled = false

[[logging]]
logger = ""
file = "/opt/go-carbon/log/go-carbon.log"
level = "info"
encoding = "mixed"
encoding-time = "iso8601"
encoding-duration = "seconds"
````

**Metric retention and aggregation schemas**
storage-schemas.conf
````
[default]
pattern = .*
retentions = 60s:30d,5m:60d,15m:365d
compressed = true
````

storage-aggregation.conf files.
````
[default]
pattern = .*
xFilesFactor = 0.5
aggregationMethod = average
````

**Additional context**
I'm looking to replace an older python based graphite that has anywhere from 5 million to 10 million metrics at any given time with a better performing one. The performance of the python based graphite at that scale is alright at best, and go-carbon is my best alternative. I've been running both the original graphite and go-carbon in parallel for the past several months to ensure that go-carbon is stable and performs well at that scale and I can confirm that it is. The only thing stopping me from cutting over to using go-carbon are some querying issues (this bug being the primary one) as well as some query performance issues that I'm working through (and might pick your brain about some other day).

**Comment by bom-d-van** [comment]
Hi @cthiel42 , thanks for the detailed report.

It's a bit of an unfortunate situation here. As in the graphite spec, escaping char `\` isn't really mentioned: https://graphite.readthedocs.io/en/latest/render_api.html#target

And in go-carbon, if you are not using trie or trigram index, it falls back to Go standard library filepath.Glob which requires an escaping char `\` before hypen `-`.

And with indexing in go-carbon, `trie-index` doesn't even support escaping at all and it's returning incorrect result, which requires a fix. `trigram-index` seems also not support it as well.

For this special issue, I would suggest you only use hypen as part of range in the square bracket `[]`, and use alternative query to filter metrics:

* replace the query with this instead: `{_,-}`: `app{_,-}prd{_,-}product.*`
* use `*` for matching, but this is less efficient, not recommended: `app*prd*product.*`

This syntax should return correct result in trie, trigram, and Go standard library search result.

> some query performance issues that I'm working through

Enabling query indexing like (`trie-index`) should helps.

**Comment by cthiel42** [comment]
My hope was that this could be fixed on the go-carbon end rather than update the query. The reason being is that I have over 200 Grafana dashboards with panels using this syntax, and I would need to build out a script to programmatically update all of them and then ensure no one continues to use the old syntax in new dashboards or updates to their dashboards, or build some functionality into carbonapi to convert syntax with `[-_]` to `[\-_]` or `{_,-}` on the fly before passing requests to carbonserver.

I just wanted to confirm that this is my only route before closing this issue, or do I have simpler options?

**Comment by bom-d-van** [comment]
Hi, it's a bit tricky on go-carbon's end. The problem is that `[-_]` isn't really legit syntax to begin with, according to the graphite documentation. 

I can try taking a look and see if it's possible to support it in `go-carbon`, but I'm sorry that I wont' be able to promise a delivery due to the special nature of this issue.

> build some functionality into carbonapi to convert syntax with [-_] to [\-_] or {_,-}

IMO, it isn't a very good idea either to support it in carbonapi. 

But if you decided to bite the bullet and update your dashboard, I encourage you to go with this syntax `{_,-}`, as it's the supported syntax for value list. `[]` is for ranges in graphite documentation, not the same as what it means in many common regular expressions.  What's more, this way you can be safe to switch on/off `trie-index` or `trigram-index`.

---

off-topic: This issue counts as [undocumented feature/bug](https://en.wikipedia.org/wiki/Undocumented_feature) in python graphite. ;)

**Comment by cthiel42** [comment]
Alright, thanks for the guidance.

I'll see about updating our syntax.

**Comment by Civil** [comment]
I would say it still might be important to support that syntax on go-carbon side if that's a low-effort work and with low priority as even if that's a undocumented feature, people still might rely on it and it's rather important to be as compatible with graphite as possible.

I can actually have a look once I'll do yet another attempt to finish the old tags support project (I hope to do another attempt in upcoming months but that heavily depends on a free time).

**Comment by cthiel42** [comment]
I've got a work around right now, but not one I think I'd want to use long term. I think if you wanted go-carbon to be a "drop in replacement", then this is an oversight (although a minor one)
